features-linux.md

```diff
@@ -0,0 +1,183 @@
+]
+```
+
+## <a name="linuxFeaturesCgroup" />Cgroup
+
+**`cgroup`** (object, OPTIONAL) represents the runtime's implementation status of cgroup managers.
+Irrelevant to the cgroup version of the host operating system.
+
+* **`v1`** (bool, OPTIONAL) represents whether the runtime supports cgroup v1.
+* **`v2`** (bool, OPTIONAL) represents whether the runtime supports cgroup v2.
+* **`systemd`** (bool, OPTIONAL) represents whether the runtime supports system-wide systemd cgroup manager.
+* **`systemdUser`** (bool, OPTIONAL) represents whether the runtime supports user-scoped systemd cgroup manager.
+
+### Example
+
+```json
+"cgroup": {
+  "v1": true,
+  "v2": true,
+  "systemd": true,
+  "systemdUser": true
+}
+```
+
+## <a name="linuxFeaturesSeccomp" />Seccomp
+
+**`seccomp`** (object, OPTIONAL) represents the runtime's implementation status of seccomp.
+Irrelevant to the kernel version of the host operating system.
+
+* **`enabled`** (bool, OPTIONAL) represents whether the runtime supports seccomp.
+* **`actions`** (array of strings, OPTIONAL) The recognized names of the seccomp actions.
+  The runtime MUST recognize the elements in this array in the [`syscalls[].action` property of the `linux.seccomp` object in `config.json`](config-linux.md#seccomp).
+* **`operators`** (array of strings, OPTIONAL) The recognized names of the seccomp operators.
+  The runtime MUST recognize the elements in this array in the [`syscalls[].args[].op` property of the `linux.seccomp` object in `config.json`](config-linux.md#seccomp).
+* **`archs`** (array of strings, OPTIONAL) The recognized names of the seccomp architectures.
+  The runtime MUST recognize the elements in this array in the [`architectures` property of the `linux.seccomp` object in `config.json`](config-linux.md#seccomp).
+
+### Example
+
+```json
+"seccomp": {
+  "enabled": true,
+  "actions": [
+    "SCMP_ACT_ALLOW",
+    "SCMP_ACT_ERRNO",
+    "SCMP_ACT_KILL",
+    "SCMP_ACT_LOG",
+    "SCMP_ACT_NOTIFY",
+    "SCMP_ACT_TRACE",
+    "SCMP_ACT_TRAP"
+  ],
+  "operators": [
+    "SCMP_CMP_EQ",
+    "SCMP_CMP_GE",
+    "SCMP_CMP_GT",
+    "SCMP_CMP_LE",
+    "SCMP_CMP_LT",
+    "SCMP_CMP_MASKED_EQ",
+    "SCMP_CMP_NE"
+  ],
+  "archs": [
+    "SCMP_ARCH_AARCH64",
+    "SCMP_ARCH_ARM",
+    "SCMP_ARCH_MIPS",
+    "SCMP_ARCH_MIPS64",
+    "SCMP_ARCH_MIPS64N32",
+    "SCMP_ARCH_MIPSEL",
+    "SCMP_ARCH_MIPSEL64",
+    "SCMP_ARCH_MIPSEL64N32",
+    "SCMP_ARCH_PPC",
+    "SCMP_ARCH_PPC64",
+    "SCMP_ARCH_PPC64LE",
+    "SCMP_ARCH_S390",
+    "SCMP_ARCH_S390X",
+    "SCMP_ARCH_X32",
+    "SCMP_ARCH_X86",
+    "SCMP_ARCH_X86_64"
+  ]
+}
+```
+
+## <a name="linuxFeaturesApparmor" />AppArmor
+
+**`apparmor`** (object, OPTIONAL) represents the runtime's implementation status of AppArmor.
+Irrelevant to the availability of AppArmor on the host operating system.
+
+* **`enabled`** (bool, OPTIONAL) represents whether the runtime supports AppArmor.
+
+### Example
+
+```json
+"apparmor": {
+  "enabled": true
```

Comment:
Not sure but should this be `supported` instead of `enabled`.
```suggestion
  "supported": true
```

Reply:
The JSON is already implemented in runc (experimentally, though), so I'm reluctant to change the field names.

IIRC, I avoided "supported" to avoid confusion with enterprise commercial "support".
An enterprise distributor may provide commercial "support" to subset of the "enabled" features.

Reply:
Oh i see. This is a really small `nit` and i don't have a strong opinion on this. :smiley: 

To me it just feels that `apparmor` is only **enabled** for a particular `spec` if `spec` defines `apparmorProfile` but by default its only **supported** by a runtime and not enabled for every `spec`.

Again i am thinking this too much and its not a big thing. :sweat_smile:

Reply:
I think `enabled` is okay, though I think that we shouldn't care what runc has developed at the moment -- `runc features` was only merged a day or two ago, so literally nobody is using at the moment (and we can change it until runc 1.2).

Reply:
> IIRC, I avoided "supported" to avoid confusion with enterprise commercial "support". An enterprise distributor may provide commercial "support" to subset of the "enabled" features.

Perhaps "available" ? (IMO "enabled" feels like a configuration toggle where most/many of these will really be capabilities that are either usable or they aren't)

Reply:
Ah, I guess https://github.com/opencontainers/runtime-spec/pull/1130/files#diff-445e6d0dc97b93a0f04d914ad7816114ddc175dc462c7eec7961921af17ff0d0R6-R7 is the context I was missing:

> The features document is irrelevant to the actual availability of the features in the host operating system. Hence, the content of the feature document SHOULD be determined on the compilation time of the runtime, not on the execution time.

I redact my suggestion ("enabled" seems fine for what this describes), but then also wonder which specific problem this functionality is solving. :see_no_evil:

Reply:
>  wonder which specific problem this functionality is solving.

Examples of specific problems solved in this PR:

- A high-level engine can't be aware whether the low-level runtime supports seccomp, so a high-level engine can't avoid "seccomp not supported" error unless the user supplies a custom configuration.
- A high-level engine can't be aware whether the low-level runtime supports `rro` mounts, so a high-level engine may accidentally have non-RRO mounts agains the user's will. This may result in a severe vulnerability.

Reply:
BTW we can have `availability` property that corresponds to actual availability of the feature on the host OS/hardware, in addition to the `enabled` property proposed in this PR, but that will be another PR.

As a maintainer of containerd/Moby and its relevant projects, I don't see necessity of having `availability` detection in the low-level runtime, though, because high-level engine implementations are already capable of detecting such feature availability on the host OS/hardware.

Reply:
> BTW we can have `availability` property that corresponds to actual availability of the feature on the host OS/hardware, in addition to the `enabled` property proposed in this PR, but that will be another PR.
> 
> As a maintainer of containerd/Moby and its relevant projects, I don't see necessity of having `availability` detection in the low-level runtime, though, because high-level engine implementations are already capable of detecting such feature availability on the host OS/hardware.

My original doubt was to replace `enabled` with `supported` or `availability` I don't think adding extra key just for `availability` would be good idea. So I agree with @AkihiroSuda we should not need extra key for this. 

+1 for keeping `enabled: true`